Use all space types for homepage stats

mrcasals opened the pull request:
#### :tophat: What? Why?
After #2221, which improved the stats count in the homepage, we realized numbers were still wrong. Only processes were being used for these calculations, so assemblies or initiatives were excluded.

This PR adds a way for participatory space manifests to define how to retrieve public spaces for a given organization. Then these values are used to calculate all the stats.

Example: The development app has 2 processes and 2 assemblies by default, each of those has all possible features. This amount to 4 proposal features, 4 meeting features, etc. For meetings, for example, we create only 2 elements per feature. This should amount to 8 meetings.

Before this PR:
![](https://i.imgur.com/bzIqq8d.png)

After this PR:
![](https://i.imgur.com/nJDFaC3.png)

Number have doubled because we always create the same amount amount of resources per feature.

#### :pushpin: Related Issues
- Related to #2221

#### :clipboard: Subtasks
None

deivid-rodriguez commented on decidim-dev/lib/decidim/dev/test/rspec_support/feature.rb:
```diff
@@ -105,6 +105,18 @@ def serialize
+
+  feature.register_stat :dummies_count_low do |features, _start_at, _end_at|
+    features.count * 200
+  end
```
Unused?

mrcasals replied:
Oh yes, looks like. I'll try to fix it this weekend, otherwise @josepjaume will have to take a look at it

deivid-rodriguez replied:
I can do it, @mrcasals!

xabier commented:
hi @mrcasals quick question... there is a new entity on the stats (that I also found at the meta.decidim portal) and I can't figure out what it is.... what is a "response / respostes" ?

mrcasals commented:
@xabier those are survey answers, added on #1526 and first released on version 0.5.0 :smile:

xabier commented:
> @xabier those are survey answers, added on #1526 and first released on version 0.5.0

It seems I haven't encountered a Decidim using surveys until recently :smile: the name is confusing, because Decidim admins respond to proposals when they are accepted or rejected. Comments also get responded.... Should it not better to name it "Completed Surveys" or "Response to Surveys"?

When defining the original issue #2003 I didn't prioritize providing stats for this feature, we had a discussion at @decidim/product to identify the most important stats. Partly because having more than 3 rows looks ugly (3 are already not very good looking) partly because nobody was using surveys at the time. I think some of the small info at the bottom is more important (meetings or supports) than Survey Responses. I suggest we put this stat smaller bellow and move up Supports or Meetings.

xabier commented:
@mrcasals , nice job!

> @xabier those are survey answers, added on #1526 and first released on version 0.5.0

It seems I haven't encountered a Decidim using surveys until recently :smile: the name is confusing, because Decidim admins respond to proposals when they are accepted or rejected. Comments also get responded.... Should it not better to name it "Completed Surveys" or "Response to Surveys"?

When defining the original issue #2003 I didn't prioritize providing stats for this feature, we had a discussion at @decidim/product to identify the most important stats. Partly because having more than 3 rows looks ugly (3 are already not very good looking) partly because nobody was using surveys at the time. I think some of the small info at the bottom is more important (meetings or supports) than Survey Responses. I suggest we put this stat smaller bellow and move up Supports or Meetings.

mrcasals commented:
@deivid-rodriguez wow, I was about to fix this PR myself, then saw you push while I was writing a comment 🤣

deivid-rodriguez commented:
Ooops, so you had already worked on it...? Sorry :|

mrcasals commented:
Nono, I was about to start! No problem at all 😄

deivid-rodriguez commented:
That's a relief :relieved: They're very simple tiny changes anyways!

deivid-rodriguez commented:
Shit, CI failed. Seems like an order dependent failure...

mrcasals commented:
Should we add this `translation missing: en.pages.home.statistics.dummies_count_high` to a locales file in `dev`?

deivid-rodriguez commented:
hehehe, that's exactly what I'm about to push. :)

mrcasals commented on decidim-dev/config/locales/en.yml:
```diff
@@ -5,6 +5,17 @@ en:
       admin:
```
LOL I thiought w didn't have any locales file in dev :sweat_smile:

josepjaume commented:
@xabier the *representation* logic for statistics is abstracted from the *extraction*. We've already gone far without carefully examining how these statistics should be displayed. I'm gonna merge this (as it actually fixes a bug) but I suggest we think carefully about how to show this information on the homepage afterwards with help from the PX guys.

josepjaume commented on decidim-core/lib/decidim/participatory_space_manifest.rb:
```diff
@@ -50,5 +50,9 @@ def seed!
       super || model_class_name.demodulize.underscore
     end
+
+    def public_spaces(&block)
```
Statistics are also used on the admin panel so this might not work there. What if we provided `participatory_spaces` and `features` and we used a `public` scope on top of them when needed? Spaces and Features are always gonna be `ActiveRecord` models so we can rely on that working in future cases.

Thoughts @mrcasals?

josepjaume replied:
Ok, after talking about it *offline*, this won't break existing functionality. But I still feel like this is a *leaky* abstraction: We're asking for a very specific subset of data. By returning *all* the participatory spaces, we could then use this to build other features on top of it, while returning only the *public* limits us in some ways.

So, what I suggest is:

1) Making the contract more generic and as for *participatory_spaces*.
2) Adding a *scope* called *public* in all participatory spaces.
3) Create a *shared example* for participatory spaces that acts as a linter so tests on spaces will fail if we change / extend this contract in the future.

@deivid-rodriguez @oriolgual @mrcasals thoughts?

deivid-rodriguez replied:
Without giving too much thought to it, what you suggest makes sense to me and it certainly makes the API more robust and stable.

mrcasals replied:
Yeah, +1 to me, but I'm out on holidays and I don't know when I'll have time to work on this :disappointed:

josepjaume replied:
No worries, this isn't an *urgent* bug, so you can happily work on it when you return xDDD

josepjaume commented on decidim-core/lib/decidim/participatory_space_manifest.rb:
```diff
@@ -50,5 +50,9 @@ def seed!
     end
+
+    def public_spaces(&block)
+      @public_resources ||= block
```
Shouldn't this be `@public_spaces` instead? Or, following my suggestion: `@participatory_spaces`.

josepjaume commented:
Ok, not merging this straight away. I'd like to tackle some things first.

xabier commented:
> @xabier the representation logic for statistics is abstracted from the extraction. We've already gone far without carefully examining how these statistics should be displayed. I'm gonna merge this (as it actually fixes a bug) but I suggest we think carefully about how to show this information on the homepage afterwards with help from the PX guys.

Good point, the roadmap already anticipates a need to review all the stats and data visualization, lets move forward with this "sub-optimal" solution now and we will review and improve this (I had in mind something like a statistics graph that grown when you scroll down to it) that is more fancy and better designed and prioritized. We have this planned by 2018Q2, meanwhile... this is ok!

mrcasals commented:
@deivid-rodriguez I'm retaking this PR, thanks for taking care of it meanwhile!

mrcasals commented:
@josepjaume OK, I've updated the code to reflect the discussion. I've added the `public_spaces` to `Decidim::Participable`, which is included by all participatory spaces. It cannot be called `public` because it's a Ruby method and redefining it makes weird things happen. 

Thoughts?

/cc @deivid-rodriguez

deivid-rodriguez commented:
@mrcasals Welcome back!

All concerns seem addressed now, I think.
